### PR TITLE
Refactor config: extract block, auth factories

### PR DESCRIPTION
**Pull request by arielshaqed**
config also holds factories.  This creates dependency loops -- which
Go's import cycle rules make hard to break.  Prevent import loops by
splitting packages that need a factory into:
1. The actual package (virtually unchanged, except to take a
   parameters object in its constructor);
1. A "params" (sub-)package, which holds parameter structs that config
   creates and factories consume;
1. A "factory" (sub-)package, which calls the appropriate config
   methods to get parameters and pass them to the constructor.

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -10,26 +10,19 @@ import (
...
-	"github.com/treeverse/lakefs/auth"
-	"github.com/treeverse/lakefs/block"
-	"github.com/treeverse/lakefs/block/local"
-	"github.com/treeverse/lakefs/block/mem"
-	s3a "github.com/treeverse/lakefs/block/s3"
-	"github.com/treeverse/lakefs/block/transient"
-	"github.com/treeverse/lakefs/db"
-	"github.com/treeverse/lakefs/logging"
+
+	auth_params "github.com/treeverse/lakefs/auth/params"
```
Comment (nopcoder) [inline config/config.go:20]: https://blog.golang.org/package-names

File: config/config.go [hunk 2]
```diff
@@ -197,64 +181,31 @@ func GetAccount(awsConfig *aws.Config) (string, error) {
...
+	}, nil
 }
 
-func (c *Config) buildLocalAdapter() (block.Adapter, error) {
+func (c *Config) GetBlockAdapterLocalParams() (block_params.Local, error) {
 	localPath := viper.GetString("blockstore.local.path")
-	location, err := homedir.Expand(localPath)
+	path, err := homedir.Expand(localPath)
 	if err != nil {
 		return nil, fmt.Errorf("could not parse blockstore location URI: %w", err)
```
Comment (nopcoder) [inline config/config.go:202]: cant return nil

File: block/params/block.go [hunk 3]
```diff
@@ -0,0 +1,19 @@
+package params
```
Comment (nopcoder) [inline block/params/block.go:1]: Not part of each block type?

Comment (arielshaqed) [reply block/params/block.go:1]: No: needed in a factory and in config.  (Of course, I _could_ still push each one out, but that would make the factory and the config have to import 3x as many params.)

File: config/config.go (hunk 1 above)
Comment (arielshaqed) [reply config/config.go]: Done -- after f2fing removed underscores.

**Comment by arielshaqed** [comment]
Deleted GCS-specific config during rebase: after F2Fing with @nopcoder GCS is already possible using the existing options, and will be changed again.

**Review by nopcoder — APPROVED** [review]
very cool

**Comment by arielshaqed** [comment]
Thanks (and especially for the merge fix), pulling!

**Pull request merged**